### PR TITLE
Add on-demand compiler pool scaling

### PR DESCRIPTION
**Pull request by fantix**
Use --compiler-pool-scaling-mode=on_demand to switch to the new mode,
which will a spawn new compiler worker process every 3 seconds if the
compiling requests keep queueing up, up to num of CPUs; after 60 seconds
without compiling requests, the compiler pool will scale down to
--compiler-pool-size with a default of 1 under on-demand mode.

### CODE REVIEW

File: edb/server/args.py [hunk 1]
```diff
@@ -94,6 +94,41 @@ class BackendCapabilitySets(NamedTuple):
...
 
 
+class CompilerPoolScalingMode(enum.StrEnum):
+    Fixed = "fixed"
+    OnDemand = "on_demand"
+
+    def __init__(self, name):
+        self.pool_class = None
+
+    def set_class(self, cls):
```
Comment (1st1) [inline edb/server/args.py:104]: I think `assign_implementation` would be less confusing. Also add a comment that this is a decorator

File: edb/server/compiler_pool/pool.py [hunk 2]
```diff
@@ -736,6 +703,201 @@ async def describe_database_restore(
...
+    async def _start(self):
+        await self._create_template_proc(retry=False)
+
+    async def _create_template_proc(self, retry=True):
+        self._template_proc_scheduled = False
+        if not self._running:
+            return
+        self._template_proc_version += 1
+        try:
+            self._template_transport = await self._create_worker_process(
```
Comment (1st1) [inline edb/server/compiler_pool/pool.py:750]: It's a bit hard to follow the changes -- are we still using fork for the "fixed" pool?

Comment (1st1) [reply edb/server/compiler_pool/pool.py:750]: This is in part because of using inheritance. Can you add comments here and there explaining the key points: where we spawn, where we fork, and how we do that?

Comment (fantix) [reply edb/server/compiler_pool/pool.py:750]: Yeah sure, np

**Comment by 1st1** [comment]
Should we merge this one?

**Review by elprans — APPROVED** [review]
:+1:
